File: src/csharp/pulse/cdm/system/environment/SEAppliedTemperature.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SEAppliedTemperature
  {
    protected SEScalarTemperature temperature;
    protected SEScalarArea        surface_area;
    protected SEScalar0To1        surface_area_fraction;
    protected eSwitch             state;

    public SEAppliedTemperature()
    {
      temperature = null;
      surface_area = null;
      surface_area_fraction = null;
      state = eSwitch.Off;
    }

    public void Copy(SEAppliedTemperature other)
    {
      if (this == other)
        return;
      Clear();
      state = other.state;
      if (other.temperature != null)
        GetTemperature().Set(other.GetTemperature());
      else if (temperature != null)
        temperature.Invalidate();
      if (other.surface_area != null)
        GetSurfaceArea().Set(other.GetSurfaceArea());
      else if (surface_area != null)
        surface_area.Invalidate();
      if (other.surface_area_fraction != null)
        GetSurfaceAreaFraction().Set(other.GetSurfaceAreaFraction());
      else if (surface_area_fraction != null)
        surface_area_fraction.Invalidate();
    }

    public void Clear()
    {
      state = eSwitch.Off;
      if (temperature != null)
        temperature.Invalidate();
      if (surface_area != null)
        surface_area.Invalidate();
      if (surface_area_fraction != null)
        surface_area_fraction.Invalidate();
    }

    public eSwitch GetState()
    {
      return state;
    }
    public void SetState(eSwitch s)
    {
      this.state = (s == eSwitch.NullSwitch) ? eSwitch.Off : s;
    }

    public bool HasTemperature()
    {
      return temperature == null ? false : temperature.IsValid();
    }
    public SEScalarTemperature GetTemperature()
    {
      if (temperature == null)
        temperature = new SEScalarTemperature();
      return temperature;
    }

    public bool HasSurfaceArea()
    {
      return surface_area == null ? false : surface_area.IsValid();
    }
    public SEScalarArea GetSurfaceArea()
    {
      if (surface_area == null)
        surface_area = new SEScalarArea();
      return surface_area;
    }

    public bool HasSurfaceAreaFraction()
    {
      return surface_area_fraction == null ? false : surface_area_fraction.IsValid();
    }
    public SEScalar0To1 GetSurfaceAreaFraction()
    {
      if (surface_area_fraction == null)
        surface_area_fraction = new SEScalar0To1();
      return surface_area_fraction;
    }

    public new string ToString()
    {
      return "Thermal Application:"
          + "\n\tState: " + GetState()
          + "\n\tAppliedTemperature: " + (this.HasTemperature() ? this.GetTemperature().ToString() : "None")
          + "\n\tAppliedSurfaceArea: " + (this.HasSurfaceArea() ? this.GetSurfaceArea().ToString() : "None")
          + "\n\tAppliedSurfaceAreaFraction: " + (this.HasSurfaceAreaFraction() ? this.GetSurfaceAreaFraction().ToString() : "None");
    }

  }
}

File: src/cpp/cdm/system/environment/SEAppliedTemperature.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
#include "stdafx.h"
#include "system/environment/SEAppliedTemperature.h"
#include "properties/SEScalarArea.h"
#include "properties/SEScalar0To1.h"
#include "properties/SEScalarPower.h"
#include "properties/SEScalarTemperature.h"

SEAppliedTemperature::SEAppliedTemperature(Logger* logger) : Loggable(logger)
{
  m_Temperature = nullptr;
  m_SurfaceArea = nullptr;
  m_SurfaceAreaFraction = nullptr;
  m_State = eSwitch::Off;
}

SEAppliedTemperature::~SEAppliedTemperature()
{
  m_State = eSwitch::Off;
  SAFE_DELETE(m_Temperature);
  SAFE_DELETE(m_SurfaceArea);
  SAFE_DELETE(m_SurfaceAreaFraction);
}

void SEAppliedTemperature::Clear()
{
  m_State = eSwitch::Off;
  INVALIDATE_PROPERTY(m_Temperature);
  INVALIDATE_PROPERTY(m_SurfaceArea);
  INVALIDATE_PROPERTY(m_SurfaceAreaFraction);
}

const SEScalar* SEAppliedTemperature::GetScalar(const std::string& name)
{
  if (name.compare("Temperature") == 0)
    return &GetTemperature();
  if (name.compare("SurfaceArea") == 0)
    return &GetSurfaceArea();
  if (name.compare("SurfaceAreaFraction") == 0)
    return &GetSurfaceAreaFraction();
  return nullptr;
}

bool SEAppliedTemperature::HasTemperature() const
{
  return m_Temperature == nullptr ? false : m_Temperature->IsValid();
}
SEScalarTemperature& SEAppliedTemperature::GetTemperature()
{
  if (m_Temperature == nullptr)
    m_Temperature = new SEScalarTemperature();
  return *m_Temperature;
}
double SEAppliedTemperature::GetTemperature(const TemperatureUnit& unit) const
{
  if (m_Temperature == nullptr)
    return SEScalar::dNaN();
  return m_Temperature->GetValue(unit);
}

bool SEAppliedTemperature::HasSurfaceArea() const
{
  return m_SurfaceArea == nullptr ? false : m_SurfaceArea->IsValid();
}
SEScalarArea& SEAppliedTemperature::GetSurfaceArea()
{
  if (m_SurfaceArea == nullptr)
    m_SurfaceArea = new SEScalarArea();
  return *m_SurfaceArea;
}
double SEAppliedTemperature::GetSurfaceArea(const AreaUnit& unit) const
{
  if (m_SurfaceArea == nullptr)
    return SEScalar::dNaN();
  return m_SurfaceArea->GetValue(unit);
}

bool SEAppliedTemperature::HasSurfaceAreaFraction() const
{
  return m_SurfaceAreaFraction == nullptr ? false : m_SurfaceAreaFraction->IsValid();
}
SEScalar0To1& SEAppliedTemperature::GetSurfaceAreaFraction()
{
  if (m_SurfaceAreaFraction == nullptr)
    m_SurfaceAreaFraction = new SEScalar0To1();
  return *m_SurfaceAreaFraction;
}
double SEAppliedTemperature::GetSurfaceAreaFraction() const
{
  if (m_SurfaceAreaFraction == nullptr)
    return SEScalar::dNaN();
  return m_SurfaceAreaFraction->GetValue();
}

eSwitch SEAppliedTemperature::GetState() const
{
  return m_State;
}
void SEAppliedTemperature::SetState(eSwitch state)
{
  m_State = (state == eSwitch::NullSwitch) ? eSwitch::Off : state;
}



void SEAppliedTemperature::ToString(std::ostream &str) const
{
  str << "Applied Temperature :";
  str << "\n\tTemperature :";         HasTemperature() ? str << *m_Temperature : str << "NaN";
  str << "\n\tSurfaceArea :";         HasSurfaceArea() ? str << *m_SurfaceArea : str << "NaN";
  str << "\n\tSurfaceAreaFraction :"; HasSurfaceAreaFraction() ? str << *m_SurfaceAreaFraction : str << "NaN";
  str << "\n\tState :" <<  eSwitch_Name(m_State);
  str << std::flush;
}